### DISCUSSION TITLE
A few extra questions on FLAML

### DISCUSSION THREAD

**Discussion by flippercy** [microsoft/FLAML#779] [General]
Hi Team FLAML:

I have a few general questions regarding FLAML:

1. Could I use a customized learner for the ensemble? For example, for a linear combination, I want to make sure all the coefficients are negative.

2. I noticed that the search of automl is much more biased when running distributedly on compute clusters, which means that it focuses much more on 'strong' learners. For example, this is the summary from my recent search:
![image](https://user-images.githubusercontent.com/51969693/198721026-9617c07c-de56-49ab-8990-3d5f15e6c7a9.png)
Obviously FLAML spent the majority of time on LearnerE, which is the optimal one; it is understandable, however, the distribution of attempts on different learners is much more balanced when running FLAML on a VM. Could you explain this phenomenon, please? Usually we use FLAML not just to look for the overall optimal solution, but also the best result per learner.

Thank you!

**Comment by sonichi** [comment]
1. Yes. Please take a look at https://microsoft.github.io/FLAML/docs/Use-Cases/Task-Oriented-AutoML#ensemble and let me know if it's what you want.
2. Is the option [blendsearch] installed when you install flaml?

**Comment by flippercy** [comment]
Hi @sonichi:

1. Yes that's what I want.
2. In the docker file flaml is installed as: 'RUN pip install flaml[blendsearch,ray]'

**Reply by sonichi** [reply]
flaml is optimized for outputting the best result given the time budget, so it's possible that it doesn't explore each learner as much as you would like to.
If you prefer tuning each learner sufficiently, how about submitting 6 flaml jobs, each tuning a single learner?
(cc @qingyun-wu for awareness)

**Reply by flippercy** [reply]
I understand this; however, the behavior of FLAML is very different (much more balanced) when running on a VM. Any idea why?

**Reply by sonichi** [reply]
The learner selection part in sequential automl uses random sampling. The learner selection part in parallel setting uses a deterministic rule. Maybe that is one reason.

**Reply by flippercy** [reply]
Make sense. Could you specify what deterministic rule is used?

**Reply by sonichi** [reply]
The search thread which leads to best estimated final score by the end is picked at each iteration. BTW, if you use `max_iter` instead of `time_budget` as the stop criterion, the result may become more balanced in terms of number of iterations per learner. If you would like to give that a try, please let me know how it works.

**Reply by flippercy** [reply]
I set time_budget = 3600*12 and max_iter=100, which means that time budget is unbounded; however, it does not impact the behavior of FLAML on compute clusters at all. The optimal learner still accounts for 90%+ of the total search.

**Reply by sonichi** [reply]
Oh, one change in the code is needed to take effect. Do you think it's worth a try? Or would you like to take the approach mentioned above to submit separate jobs for different learners?

**Reply by flippercy** [reply]
There is no hurry because for small projects we can always just use a large VM; however, it would be great that if this issue can be fixed in the future. From the ROA perspective it is better to let the algorithm make the decision on how to allocate resources to all the learners. 

Thank you!

**Reply by sonichi** [reply]
I created an issue #789. Would you please test it after it's fixed? Welcome to join our discord to discuss further.

**Reply by flippercy** [reply]
Thanks @sonichi . Back to question #1 on ensemble; my team created the following codes to use ElasticNet as a customized learner for the ensemble; however, it does not work. Could you tell me what the issues are, please?

from flaml.model import SKLearnEstimator
from sklearn.linear_model import ElasticNet

class MyElasticNetClassifier(SKLearnEstimator):

    def __init__(self, positive = True, **params):

        super().__init__(task, **params)

        self.estimator_class = ElasticNet

        # convert to int for integer hyperparameters
        self.params = {
            'random_state': params['random_state'] if 'random_state' in params else randomseed,
            'positive': params['positive'] if 'positive' in params else True,   
            'alpha': params['alpha'], 
       }   

    @classmethod
    def search_space(cls, data_size, task):
        space = {
            "alpha": {"domain": tune.uniform(lower=0, upper=1), "init_value": 0.5},
        }
        return space

automl.add_learner(learner_name = 'MyElasticNet', learner_class = MyElasticNetClassifier)

settings = {
    "ensemble": {"final_estimator": MyElasticNetClassifier,
                "passthrough": False,
                }
}

Thank you!

**Reply by sonichi** [reply]
The "final_estimator" expects an instance of the estimator instead of the class. Also, the estimator doesn't need to be registered using `add_learner`. It can be any scikit-learn style estimator.

**Reply by flippercy** [reply]
So I guess I can simply use:

"final_estimator": ElasticNet(positive=True),

correct?

**Reply by sonichi** [reply]
Yes, assuming `ElasticNet(positive=True)` is a valid sklearn classifier.
